### PR TITLE
[#10959] Add endpoint to fetch pagination metadata

### PR DESCRIPTION
**Pull request by daongochieu2810**
Part of #10959

Endpoint to fetch the total number of responses for a question in order to prepare pagination calls.
Path: ```/webapi/pagination/metadata```
Params: course ID, instructor email, question ID
Response: number of submitted responses, question ID

### CODE REVIEW

File: src/main/java/teammates/common/util/Const.java [hunk 1]
```diff
@@ -265,6 +265,7 @@ private Const() {
         public static final String INSTRUCTOR = URI_PREFIX + "/instructor";
         public static final String INSTRUCTOR_PRIVILEGE = URI_PREFIX + "/instructor/privilege";
         public static final String RESULT = URI_PREFIX + "/result";
+        public static final String PAGINATION_META_DATA = URI_PREFIX + "/pagination/metadata";
```
Comment (moziliar) [inline src/main/java/teammates/common/util/Const.java:268]: Does this endpoint have to reside in a separate routing group? If so, what other calls could there possibly be in the same `/pagination/` routing group?

File: src/main/java/teammates/logic/api/Logic.java [hunk 2]
```diff
@@ -1387,4 +1387,8 @@ public int getNumOfGeneratedChoicesForParticipantType(String courseId, FeedbackP
         return feedbackQuestionsLogic.getNumOfGeneratedChoicesForParticipantType(courseId, generateOptionsFor);
     }
 
+    public int getSubmittedResponseCountForQuestion(String questionId) {
```
Comment (moziliar) [inline src/main/java/teammates/logic/api/Logic.java:1390]: I have some doubt over the granularity of the metadata fetch:
* Is per question response count really useful? (maybe one could just use the question result call to get all the response and count the length in an inexpensive way)
* I think you could actually make it course/session wide to improve the usefulness per call.

File: src/main/java/teammates/storage/api/FeedbackResponsesDb.java [hunk 3]
```diff
@@ -399,6 +399,15 @@ private FeedbackResponse getFeedbackResponseEntity(String feedbackResponseId) {
                 .list();
     }
 
+    /**
+     * Get number of submitted responses for a question.
+     */
+    public int getSubmittedResponseCountForQuestion(String questionId) {
+        return load()
+                .filter("feedbackQuestionId =", questionId)
```
Comment (moziliar) [inline src/main/java/teammates/storage/api/FeedbackResponsesDb.java:407]: Can check if this filter reads into the document? And what cost will this bring?
Can use the mock script to populate some data and test the cost of this operation.

File: src/main/java/teammates/ui/webapi/GetPaginationMetaDataAction.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import teammates.common.exception.UnauthorizedAccessException;
+import teammates.common.util.Const;
+import teammates.ui.output.PaginationMetaData;
+
+/**
+ * Get pagination metadata.
+ */
+public class GetPaginationMetaDataAction extends Action {
+
+    private static final String MESSAGE_NOT_INSTRUCTOR_ACCOUNT = "You did not login as an instructor,"
```
Comment (moziliar) [inline src/main/java/teammates/ui/webapi/GetPaginationMetaDataAction.java:13]: message doesn't seem to match the intent

File: src/main/java/teammates/ui/webapi/GetPaginationMetaDataAction.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            + " so you cannot view your profile";
+    private static final String MESSAGE_INSTRUCTOR_NOT_FOUND = "The instructor is not in the course you are given,"
```
Comment (moziliar) [inline src/main/java/teammates/ui/webapi/GetPaginationMetaDataAction.java:15]: ditto

File: src/main/java/teammates/ui/webapi/GetPaginationMetaDataAction.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        }
+        gateKeeper.verifyAccessible(instructor, logic.getCourse(courseId),
+                Const.InstructorPermissions.CAN_VIEW_SESSION_IN_SECTIONS);
+    }
+
+    @Override
+    ActionResult execute() {
+        String feedbackQuestionId = getRequestParamValue(Const.ParamsNames.FEEDBACK_QUESTION_ID);
+
+        int submittedResponseCount = logic.getSubmittedResponseCountForQuestion(feedbackQuestionId);
```
Comment (moziliar) [inline src/main/java/teammates/ui/webapi/GetPaginationMetaDataAction.java:47]: what happened if the question does not exist? This doesn't suggest any of those errors.

**Review by moziliar — CHANGES_REQUESTED** [review]
Maybe we should discuss the entire workflow, then look into the exact implementation details. This might be an expensive call that could create noise in monitoring.

**Comment by rrtheonlyone** [comment]
I think we can close this PR - are we planning on doing anything further with this? @daongochieu2810 @moziliar

**Comment by daongochieu2810** [comment]
I dont think we need to be really detailed on pagination, so yes we should close this PR

**Pull request closed without merging**